src/main/scala/org/broadinstitute/dsde/firecloud/service/TSVFileSupport.scala

```diff
@@ -143,10 +146,31 @@ trait TSVFileSupport {
       val nameEntry = "attributeName" -> AttributeString(attributeName)
       def valEntry( attr: Attribute ) = "addUpdateAttribute" -> attr
       refTypeOpt match {
-        case Some(refType) => Map(upsertAttrOperation,nameEntry,valEntry(AttributeEntityReference(refType,value)))
+        case Some(refType) => Seq(Map(upsertAttrOperation,nameEntry,valEntry(AttributeEntityReference(refType,value))))
         case None => value match {
-          case "__DELETE__" => Map(removeAttrOperation,nameEntry)
-          case _ => Map(upsertAttrOperation,nameEntry,valEntry(AttributeString(value)))
+          case "__DELETE__" => Seq(Map(removeAttrOperation,nameEntry))
+          case value if modelSchema.isAttributeArray(value) => {
+            val listElements = value.parseJson.convertTo[JsArray].elements.toList
+            val listNameEntry = "attributeListName" -> AttributeString(attributeName)
+            def listValEntry( attr: Attribute ) = "newMember" -> attr
+
+            listElements match {
+              case elements if elements.forall(_.isInstanceOf[JsString]) =>
+                val elementsTyped: List[JsString] = elements.asInstanceOf[List[JsString]]
```

Comment:
is there a cleaner way to do this than `asInstanceOf`ing?

i tried coercing the types by making the case statements look something like: 

`case elements: List[JsString] if elements.forall(_.isInstanceOf[JsString]) =>`

but it didn't work because it causes premature casts during pattern matching...

Reply:
What about something like this?
```
          case value if modelSchema.isAttributeArray(value) =>
            val listElements = value.parseJson.convertTo[JsArray].elements.toList

            if (listElements.isEmpty) {
              // TODO: check for zero-element array and skip other processing, what should this return?
              Seq()
            } else {
              // validate that all elements are the same type
              val elClasses = listElements.map(_.getClass).toSet
              // TODO: this fails on list-of-booleans, because it resolves to Set(JsTrue, JsFalse)
              if (elClasses.size > 1) {
                // probably should remove the output of what classes we actually found
                throw new FireCloudException(s"Mixed-type entity attribute lists are not supported. Found: ${elClasses.map(_.getName).mkString(",")}")
              }

              val listNameEntry = "attributeListName" -> AttributeString(attributeName)
              def listValEntry( attr: Attribute ) = "newMember" -> attr

              // since all elements are the same type, we can just match on each one, trusting they'll be the same
              listElements.map {
                case jsstr:JsString => Map(addListMemberOperation, listNameEntry, listValEntry(AttributeString(jsstr.value)))
                case jsnum:JsNumber => Map(addListMemberOperation, listNameEntry, listValEntry(AttributeNumber(jsnum.value)))
                case jsbool:JsBoolean => Map(addListMemberOperation, listNameEntry, listValEntry(AttributeBoolean(jsbool.value)))
                case _ => throw new FireCloudException("Only string, number, and boolean attribute lists are supported")
              }
            }
```

Note the important TODO inline; this fails the list-of-booleans unit test … but I think you get where I'm going with it and maybe this will help?

Reply:
I'm not sure that I find this more readable than the current approach. Is casting with `asInstanceOf` a dangerous operation to do if I've already checked for types using `isInstanceOf`? I generally try to avoid it, but if it's not a meltdown risk, then I'd lean towards keeping it as-is

re: the TODO, I will add some logic to better handle empty lists